cpp/oneapi/dal/algo/pca/backend/gpu/train_kernel_cov_dpc.cpp

```diff
@@ -48,24 +48,70 @@ auto compute_sums(sycl::queue& q,
 }
 
 template <typename Float>
-auto compute_correlation(sycl::queue& q,
-                         const pr::ndview<Float, 2>& data,
-                         const pr::ndview<Float, 1>& sums,
-                         const dal::backend::event_vector& deps = {}) {
-    ONEDAL_PROFILER_TASK(compute_correlation, q);
-    ONEDAL_ASSERT(data.get_dimension(1) == sums.get_dimension(0));
+auto compute_means(sycl::queue& q,
+                   std::int64_t row_count,
+                   const pr::ndview<Float, 1>& sums,
+                   const bk::event_vector& deps = {}) {
+    ONEDAL_PROFILER_TASK(compute_means, q);
+    ONEDAL_ASSERT(sums.has_data());
+    ONEDAL_ASSERT(sums.get_dimension(0) > 0);
+
+    const std::int64_t column_count = sums.get_dimension(0);
+    auto means = pr::ndarray<Float, 1>::empty(q, { column_count }, alloc::device);
+    auto means_event = pr::means(q, row_count, sums, means, deps);
+    return std::make_tuple(means, means_event);
+}
+
+template <typename Float>
+auto compute_variances(sycl::queue& q,
+                       const pr::ndview<Float, 2>& cov,
+                       const bk::event_vector& deps = {}) {
+    ONEDAL_PROFILER_TASK(compute_vars, q);
+    auto column_count = cov.get_dimension(0);
+    auto vars = pr::ndarray<Float, 1>::empty(q, { column_count }, alloc::device);
+    auto vars_event = pr::variances(q, cov, vars, deps);
+    return std::make_tuple(vars, vars_event);
+}
+
+template <typename Float>
+auto compute_covariance(sycl::queue& q,
+                        std::int64_t row_count,
+                        const pr::ndview<Float, 2>& data,
+                        const pr::ndarray<Float, 1>& sums,
+                        const bk::event_vector& deps = {}) {
+    ONEDAL_PROFILER_TASK(compute_covariance, q);
+    ONEDAL_ASSERT(sums.has_data());
+    ONEDAL_ASSERT(xtx.has_data());
 
     const std::int64_t column_count = data.get_dimension(1);
-    auto corr =
-        pr::ndarray<Float, 2>::empty(q, { column_count, column_count }, sycl::usm::alloc::device);
-    auto means = pr::ndarray<Float, 1>::empty(q, { column_count }, sycl::usm::alloc::device);
-    auto vars = pr::ndarray<Float, 1>::empty(q, { column_count }, sycl::usm::alloc::device);
-    auto tmp = pr::ndarray<Float, 1>::empty(q, { column_count }, sycl::usm::alloc::device);
-    auto gemm_event = gemm(q, data.t(), data, corr, Float(1), Float(0), deps);
-    auto corr_event = pr::correlation(q, data, sums, means, corr, vars, tmp, { gemm_event });
-
-    auto smart_event = dal::backend::smart_event{ corr_event }.attach(tmp);
-    return std::make_tuple(corr, means, vars, smart_event);
+
+    auto cov = pr::ndarray<Float, 2>::empty(q, { column_count, column_count }, alloc::device);
+
+    auto gemm_event = gemm(q, data.t(), data, cov, Float(1), Float(0), deps);
+
+    auto cov_event = pr::covariance(q, row_count, sums, cov, { gemm_event });
+    return std::make_tuple(cov, cov_event);
```

Comment:
I am expecting that you need to wait for cov_event here because of compiler issue with incorrect processing of indirect memory deps (in this certain case it is cov ndarray).  Please check in nightly and benchmarks that it works well without wait here.